contrib/lib/src/compression/mod.rs

```diff
@@ -0,0 +1,111 @@
+//! `Compression` fairing and `Compressed` responder to automatically and
+//! on demand respectively compressing responses.
+mod fairing;
+mod responder;
+
+pub use self::fairing::Compression;
+pub use self::responder::Compressed;
+
+crate use self::fairing::Context;
+use rocket::http::hyper::header::{ContentEncoding, Encoding};
+use rocket::{Request, Response};
+use std::io::Read;
+
+#[cfg(feature = "brotli_compression")]
+use brotli::enc::backward_references::BrotliEncoderMode;
+
+#[cfg(feature = "gzip_compression")]
+use flate2::read::GzEncoder;
+
+crate struct CompressionUtils;
+
+impl CompressionUtils {
+    fn accepts_encoding(request: &Request, encoding: &str) -> bool {
+        request
+            .headers()
+            .get("Accept-Encoding")
+            .flat_map(|accept| accept.split(","))
+            .map(|accept| accept.trim())
+            .any(|accept| accept == encoding)
+    }
+
+    fn already_encoded(response: &Response) -> bool {
+        response.headers().get("Content-Encoding").next().is_some()
+    }
+
+    fn set_body_and_encoding<'r, B: Read + 'r>(
+        response: &mut Response<'r>,
+        body: B,
+        encoding: Encoding,
+    ) {
+        response.set_header(ContentEncoding(vec![encoding]));
+        response.set_streamed_body(body);
+    }
+
+    fn skip_encoding(
+        content_type: &Option<rocket::http::ContentType>,
+        context: &rocket::State<Context>,
+    ) -> bool {
+        match content_type {
+            Some(content_type) => context.exclusions.iter().any(|exc_media_type| {
+                if exc_media_type.sub() == "*" {
+                    *exc_media_type.top() == *content_type.top()
+                } else {
+                    *exc_media_type == *content_type.media_type()
+                }
+            }),
+            None => false,
+        }
+    }
+
+    fn compress_response(request: &Request, response: &mut Response, respect_excludes: bool) {
+        if CompressionUtils::already_encoded(response) {
+            return;
+        }
+
+        let content_type = response.content_type();
+
+        if respect_excludes {
+            let context = request
+                .guard::<::rocket::State<Context>>()
+                .expect("Compression Context registered in on_attach");
+
+            if CompressionUtils::skip_encoding(&content_type, &context) {
+                return;
+            }
+        }
+
+        // Compression is done when the request accepts brotli or gzip encoding
+        // and the corresponding feature is enabled
+        if cfg!(feature = "brotli_compression") && CompressionUtils::accepts_encoding(request, "br")
```

Comment:
Could be nice to check gzip and brotli feature upon the fairing's on_attach and save in a variable which compression to use.
A user might not see what he is missing, or think his responses are compressed but are not.
Plus it would save some instructions on every single requests.

Reply:
The only checks are 1) "is brotli enabled" and 2) "did this request say it supports brotli". 1 is done at compile time with the `cfg!`, and 2 must be checked again for every request.

Reply:
Thanks, I forgot cfg! was compile time only,

Great Pr btw

Reply:
Still half my point was that the cfg! check should also be done in on_attach so that if none of the required features are set the user can get an error instead of discovering at runtime that something is amiss, given the user has not read properly the documentation.

Reply:
In `lib.rs` there is these lines, so if none of compression features are enabled a compile error is raised:

```
#[cfg(any(feature = "brotli_compression", feature = "gzip_compression"))]
pub mod compression;
```

Regards.

Reply:
Ok didn't know about this.